db/version_set.cc

```diff
@@ -1100,9 +1107,58 @@ class LevelIterator final : public InternalIterator {
   }
 
   void Prepare(const std::vector<ScanOptions>* scan_opts) override {
+    // We assume here that scan_opts is sorted such that
+    // scan_opts[0].range.start < scan_opts[1].range.start, and non overlapping
     scan_opts_ = scan_opts;
-    if (file_iter_.iter()) {
-      file_iter_.Prepare(scan_opts_);
+    if (scan_opts_ != nullptr) {
+      std::unordered_map<size_t, std::vector<ScanOptions>> file_to_scan_opts;
+      for (size_t k = 0; k < scan_opts_->size(); k++) {
+        const ScanOptions& opt = scan_opts_->at(k);
+        auto start = opt.range.start;
+        auto end = opt.range.limit;
+
+        if (!start.has_value()) {
+          continue;
+        }
+        InternalKey istart(start.value(), kMaxSequenceNumber,
+                           kValueTypeForSeek);
+
+        // We can capture this case in the future, but for now lets skip this.
+        if (!end.has_value()) {
+          continue;
+        }
+        InternalKey iend(end.value(), kMaxSequenceNumber, kValueTypeForSeek);
+
+        // TODO: This needs to be optimized, right now we iterate twice, which
+        // we dont need to. We can do this in N rather than 2N.
+        size_t fstart = FindFile(icomparator_, *flevel_, istart.Encode());
+        size_t fend = FindFile(icomparator_, *flevel_, iend.Encode());
+
+        // We need to check the relevant cases
+        // Cases:
+        // 1. [  S        E  ]
+        // 2. [  S  ]  [  E  ]
+        // 3. [  S  ] ...... [  E  ]
+        if ((fstart < flevel_->num_files) && fstart == fend &&
+            prepared_iters[fstart] != nullptr) {
+          // Case 1
+          file_to_scan_opts[fstart].push_back(opt);
+        } else {
+          // Case 2 and 3
+          for (auto i = fstart; i <= fend; i++) {
+            if (i < flevel_->num_files) {
+              file_to_scan_opts[i].push_back(opt);
+            }
+          }
+        }
+      }
+
+      // We now have our list of intersections
+      for (const auto& [k, v] : file_to_scan_opts) {
+        auto iter = NewFileIterator(k);
+        prepared_iters[k] = iter;
+        iter->Prepare(&v);
```

Comment:
It seems the vector of ScanOptions will be destructed at the end of Prepare(). We should either make a copy of the ScanOptions inside BlockBasedTableIterator or keep this vector alive throughout the iterator lifecycle.